Refs #31300 -- Add function-based virtual fields

jpnauta opened the pull request:
As proposed in [this forum](https://groups.google.com/g/django-developers/c/9Mf7YqDA4bg/m/wb07E71oAAAJ), [this forum](https://groups.google.com/g/django-developers/c/ADSuUUuZp3Q/m/eZGYZv74AQAJ) and [this ticket](https://code.djangoproject.com/ticket/31300). Kudos to Paolo for the thorough research on backend support for this feature.

Code samples:

```
from django.db import models

# Basic usage
class MyModel(models.Model):
    a = models.IntegerField()
    b = models.IntegerField()
    a_plus_b = models.GeneratedField(expression=F("a") + F("b"))

# Option to customize how field is persisted
class MyModel(models.Model):
    name = models.CharField(max_length=10)
    lower_name = models.GeneratedField(
        expression=Lower("name"),
        output_field=TextField(
            db_collation="fr_FR"
        ),
    )

# Option to not store as column in database (not supported in Postgres)
class MyModel(models.Model):
    name = models.CharField(max_length=10)
    lower_name = models.GeneratedField(
        expression=Lower("name", output_field=TextField()), db_persist=False
    )
```


Implementation notes:
- The feature is implemented for Postgres, SQLite, and MySQL/MariaDB
- This feature has limitations on MySQL/MariaDB and older versions of SQLite because the generated field value won't be set on `save()` since `INSERT...RETURNING` isn't implemented for this DB backend
- Postgres does not support `GENERATED ALWAYS ... VIRTUAL` fields 
- SQLite doesn't support constants being injected into the `CREATE TABLE` statement
- It should be easy to implement this feature for Oracle, I just didn't bother quite yet since it's tedious to run Oracle locally

Please let me know what you think! This is a tentative first draft, so feedback and ideas are greatly appreciated 😄

jpnauta commented on django/db/backends/base/schema.py:
```diff
@@ -826,6 +853,11 @@ def alter_field(self, model, old_field, new_field, strict=False):
                 "(you cannot alter to or from M2M fields, or add or remove "
                 "through= on M2M fields)" % (old_field, new_field)
             )
+        elif old_computed_parameters != new_computed_parameters:
+            raise ValueError(
+                "Modifying ComputedFields is not supported - the field %s must "
+                "be dropped and recreated." % new_field
+            )
```
AFAICT Postgres does not support the concept of modifying `GENERATED ALWAYS` columns, so it felt easiest to disable the ability to modify these columns altogether

jpnauta commented on django/db/backends/sqlite3/features.py:
```diff
@@ -42,6 +42,9 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_update_conflicts = Database.sqlite_version_info >= (3, 24, 0)
     supports_update_conflicts_with_target = supports_update_conflicts
+    supports_computed_field = True
+    supports_computed_field_params = False
```
SQLite does not allow params to be passed into the `CREATE TABLE` SQL statement

jpnauta commented on django/db/backends/postgresql/features.py:
```diff
@@ -70,6 +70,9 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_covering_indexes = True
+    supports_computed_field = True
+    supports_computed_field_params = True
+    supports_computed_field_virtual = False
```
Postgres only supports `GENERATED ALWAYS ... STORED` columns, not `GENERATED ALWAYS ... VIRTUAL`

jpnauta commented on django/db/backends/sqlite3/schema.py:
```diff
@@ -233,7 +233,10 @@ def is_self_referential(f):
             body[create_field.name] = create_field
             # Choose a default and insert it into the copy map
-            if not create_field.many_to_many and create_field.concrete:
+            if (
+                not (create_field.many_to_many or create_field.computed)
+                and create_field.concrete
+            ):
```
When cloning a SQLite database to add a new column, computed columns should not be included or they will cause an error

jpnauta commented on django/db/models/query.py:
```diff
@@ -1792,6 +1792,9 @@ def _insert(
         Insert a new record for the given model. This provides an interface to
         the InsertQuery class and is how Model.save() is implemented.
         """
+        if self.model._meta.computed_fields and fields:
+            fields = [f for f in fields if f not in self.model._meta.computed_fields]
```
When performing a `save()`, `create()` or `bulk_create()`, values for compute fields should not be specified

adamchainz replied:
Using `in <list>` leads to O(n²) runtime, can you not use the flag on the field?

```suggestion
        if self.model._meta.computed_fields and fields:
            fields = [f for f in fields if not f.computed]
```

jpnauta commented on docs/ref/checks.txt:
```diff
@@ -234,6 +234,9 @@ Model fields
 * **fields.W907**: ``django.contrib.postgres.fields.CITextField`` is
   deprecated. Support for it (except in historical migrations) will be removed
   in Django 5.1.
+* **fields.E220**: ``<database>`` does not support ``ComputedField``\s.
+* **fields.E221**: ``<database>`` does not support ``ComputedField`` with constant parameters\s.
+* **fields.E222**: ``<database>`` does not support ``ComputedField`` with no persistence\s.
```
This is my first P/R, so I am not sure what the protocol is for adding new error checks. Any advise would be greatly appreciated!

jpnauta commented on docs/ref/checks.txt:
```diff
@@ -234,6 +234,9 @@ Model fields
 * **fields.W907**: ``django.contrib.postgres.fields.CITextField`` is
   deprecated. Support for it (except in historical migrations) will be removed
   in Django 5.1.
+* **fields.E220**: ``<database>`` does not support ``ComputedField``\s.
+* **fields.E221**: ``<database>`` does not support ``ComputedField``\s with constant parameters.
+* **fields.E222**: ``<database>`` does not support ``ComputedField``\s with no persistence.
```
This is my first P/R, so I am not sure what the protocol is for adding new error checks. Any advise would be greatly appreciated!

pauloxnet commented:
Jeremy I'm glad to see your PR based on my proposal.

I'll try to review it in few days.

Have see the comment from Simon in the mailing list about a nulla le base field to use custom expression?
https://groups.google.com/g/django-developers/c/9Mf7YqDA4bg/m/wb07E71oAAAJ

jpnauta commented:
> Jeremy I'm glad to see your PR based on my proposal.
> 
> I'll try to review it in few days.
> 
> Have see the comment from Simon in the mailing list about a nulla le base field to use custom expression? https://groups.google.com/g/django-developers/c/9Mf7YqDA4bg/m/wb07E71oAAAJ

Thanks Paolo, I appreciate it! Yes, I have included Simon's suggestion to have an optional `output_field`, which by default is calculated using the given expression. I have added usages example to the P/R description accordingly :smile:

jpnauta commented:
Hi @pauloxnet any thoughts on this? Just wanted to check-in 😄

rmortes commented:
Hey! I've been waiting for this functionality for months, and I've just seen it's being developed right now. Great work!!

I have a question about this feature: Would it support subqueries? Say I'm building a bank, and I take a `Snapshot` of my `Users` money daily. This `Snapshot` has a float `money` field and a FK to User. Would it be possible to add a new computed field `delta` with the value `F("money") - Snapshot.objects.filter(user=F("user")).order_by("-created_date").first().money`?

Thanks!

jpnauta commented:
Hi @rmortes, thanks for the interest in this P/R! Sadly I suspect your use case will indeed not work. As explained in [Postgres' docs](https://www.postgresql.org/docs/current/ddl-generated-columns.html), *the generation expression can only use immutable functions and cannot use subqueries or reference anything other than the current row in any way*. I'm not sure of the exact limitations for other DBs, but I suspect they are similar to Postgres.

pauloxnet commented:
> Hi @pauloxnet any thoughts on this? Just wanted to check-in smile

Sorry for the late reply. I'll try to take a proper look to this PR before the weekend.

jpnauta commented:
@pauloxnet (or anyone else) are you willing to review this?

adamchainz commented on tests/invalid_models_tests/test_ordinary_fields.py:
```diff
@@ -1057,3 +1058,61 @@ class Meta:
+    @skipIfDBFeature("supports_computed_field_params")
+    def test_params_not_supported(self):
+        class Model(models.Model):
+            field = models.ComputedField(Value("value", output_field=TextField()))
```
You shouldn't need `output_field` here, as `Value` can determine it automatically:

```suggestion
            field = models.ComputedField(Value("value"))
```

adamchainz commented on django/db/models/fields/computed.py:
```diff
@@ -0,0 +1,131 @@
+        databases = kwargs.get("databases") or []
+        return [
+            *super().check(**kwargs),
+            *self._check_sql_expression(databases),
+            *self._check_persistence(databases),
+        ]
```
IMO we could merge these into the `check()` function here, it would save a lot of temporary lists and looping over the databases list.

jpnauta replied:
I would tend to disagree here - every other `Field` in the django project follows this convention, so I recommend we do the same.

<img width="458" alt="Screen Shot 2023-03-25 at 6 23 50 PM" src="https://user-images.githubusercontent.com/3439947/227748606-354d09ca-b9b0-458b-a7c7-bd1f3e8b7cfd.png">
<img width="564" alt="Screen Shot 2023-03-25 at 6 23 21 PM" src="https://user-images.githubusercontent.com/3439947/227748595-3c2d4b44-740b-4639-9f41-01b5c94c748f.png">
<img width="623" alt="Screen Shot 2023-03-25 at 6 24 11 PM" src="https://user-images.githubusercontent.com/3439947/227748617-13538d62-b2cf-479d-8fcd-ba0ad7fef2d3.png">

adamchainz commented on django/db/models/fields/computed.py:
```diff
@@ -0,0 +1,131 @@
+        self._output_field = output_field
+        self.persisted = persisted
+
+        kwargs["editable"] = False
+        kwargs["blank"] = True
+        kwargs["default"] = None
```
I think we should error if any of these are passed in `**kwargs`

adamchainz commented on django/db/models/fields/computed.py:
```diff
@@ -0,0 +1,131 @@
+    _resolved_expression = None
+    output_field = None
+
+    def __init__(self, expression, output_field=None, persisted=True, *args, **kwargs):
```
`persisted` is a slightly odd param name, since it’s in the past tense. We don’t have any other past tense names that I can find in other fields.

I think `persist` would be a better name, or maybe `db_persist`, since it's a database-level thing, like `db_column` ?

Also taking `args` allows duplicate arguments for e.g. `editable`. Let's make this class [keyword-only](https://lukeplant.me.uk/blog/posts/keyword-only-arguments-in-python/):

```suggestion
    def __init__(self, *, expression, output_field=None, persisted=True, **kwargs):
```

(`Field` really should be keyword-only, but it's too late to fix that.)

jpnauta replied:
Changed to `db_persist`

adamchainz commented:
Also I would prefer we used the name `GeneratedField`, sticking with the SQL syntax, as previously discussed on the mailing list. I don’t see why we would copy SQL Alchemy, and “computed” isn’t any clearer IMO.

And I would like to say thanks for the work so far!

jpnauta commented:
Thank you for the helpful feedback @pauloxnet and @adamchainz! I believe I have addressed all your comments - please review once more and let me know if I missed anything.

pauloxnet commented:
As I wrote before the change looks good to me, I just think it's missing a section added in the model fields documentation, together with other fields.

jpnauta commented:
Thanks @pauloxnet, missed that part. I have added field documentation now, let me know if there's any other sections I missed.

pauloxnet commented:
> Thanks @pauloxnet, missed that part. I have added field documentation now, let me know if there's any other sections I missed.

Thanks.

I've unset the "Patch needs improvement" flag from the ticket.

Now we await the reviews of other Django developers.

felixxm commented on docs/ref/models/fields.txt:
```diff
@@ -1150,6 +1150,44 @@ when :attr:`~django.forms.Field.localize` is ``False`` or
 
+``GeneratedField``
+------------------
+
```
`versionadded` annotation is missing.

felixxm commented on django/db/backends/base/features.py:
```diff
@@ -339,6 +339,13 @@ class BaseDatabaseFeatures:
     supports_comments_inline = False
 
+    # Does the backend support generated columns?
+    supports_generated_columns = False
+    # Does the backend allow parameters in generated column definitions?
+    supports_generated_columns_params = False
+    # Does the backend support virtual generated columns?
+    supports_virtual_generated_columns = False
```
Base feature flags don't match feature flags in database backends.

felixxm commented on tests/invalid_models_tests/test_ordinary_fields.py:
```diff
@@ -1057,3 +1058,62 @@ class Meta:
+
+@isolate_apps("invalid_models_tests")
+class GeneratedFieldTests(TestCase):
+    @skipIfDBFeature("supports_generated_field")
+    def test_field_not_supported(self):
+        class Model(models.Model):
+            name = models.IntegerField()
+            field = models.GeneratedField(expression=F("name"))
+
+        self.assertEqual(
+            Model._meta.get_field("field").check(databases={"default"}),
+            [
+                Error(
+                    "%s does not support GeneratedFields." % connection.display_name,
+                    obj=Model._meta.get_field("field"),
+                    id="fields.E220",
+                ),
+            ],
+        )
```
This should be checked for both cases:
```suggestion
    def test_field_not_supported(self):
        class Model(models.Model):
            name = models.IntegerField()
            field = models.GeneratedField(expression=F("name"))

        self.assertEqual(
            Model._meta.get_field("field").check(databases={"default"}),
            [
                Error(
                    "%s does not support GeneratedFields." % connection.display_name,
                    obj=Model._meta.get_field("field"),
                    id="fields.E220",
                ),
            ] if not connection.features.supports_generated_field else [],
        )
```
The same changes are needed for other feature flags. Also, we should check that `required_db_features` is respected (see for example `test_db_comment_required_db_features`).

felixxm commented on django/db/backends/sqlite3/features.py:
```diff
@@ -42,6 +42,9 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_json_field_contains = False
     supports_update_conflicts = Database.sqlite_version_info >= (3, 24, 0)
     supports_update_conflicts_with_target = supports_update_conflicts
+    supports_generated_field = True
+    supports_generated_field_params = False
+    supports_generated_field_virtual = True
```
Is it true for all supported versions of SQLite (i.e. 3.21+)? I tried with SQLite 3.26 and tests crash:
```
...
django.db.utils.OperationalError: near "AS": syntax error
```

pauloxnet replied:
As I already reported in the developer mailing list [1], SQLite added support for Generated Columns in version 3.31.0 (2020-01-22) [2].

- [1] https://groups.google.com/g/django-developers/c/9Mf7YqDA4bg/m/D5a5jUCACwAJ
- [2] https://www.sqlite.org/releaselog/3_31_0.html

pauloxnet commented:
> ... Support for virtual fields on Oracle is missing.

Reiterating my very little knowledge of Oracle, I had investigated a year ago there seemed to be support for "Virtual Columns" from Oracle Database 11g Release 1
https://oracle-base.com/articles/11g/virtual-columns-11gr1

adamchainz commented on docs/ref/models/fields.txt:
```diff
@@ -1150,6 +1150,44 @@ when :attr:`~django.forms.Field.localize` is ``False`` or
+    An :class:`Expression` used by the database to automatically set the field
+    value whenever the model is changed.
+
+    Several restrictions apply to generated columns, depending on the database.
+    A :class:`GeneratedField` should be immutable and only reference
+    concrete columns inside the model itself. References to other models' fields
+    is not supported. See your database's documentation for more information.
```
This paragraph was mixing the concepts of columns and fields. Plus I think "deterministic" is the right word, rather than "immutable" (the input fields into the expression are allowed to mutate). My attempt to improve:

```suggestion
    The expressions should be deterministic and only reference fields within
    the model (in one database table). Database backends can impose further
    restrictions.
```

The "(in one database table)" is because a model can span multiple tables, in the case of multi-table inheritance.

jpnauta replied:
I pulled "immutable" from the Postgres docs. That said, "deterministic" works for me and I'm good with these changes.

adamchainz commented on docs/ref/models/fields.txt:
```diff
@@ -1150,6 +1150,44 @@ when :attr:`~django.forms.Field.localize` is ``False`` or
+
+    Default is ``True``. PostgreSQL only supports persisted columns.
+
+``GeneratedField`` is supported on MariaDB, MySQL, PostgreSQL, and SQLite. SQLite
+does not support parameters in generated column definitions.
```
I think this should go before the attribute description?

pauloxnet replied:
I agree

adamchainz commented on docs/releases/5.0.txt:
```diff
@@ -279,6 +279,8 @@ Models
   :ref:`Choices classes <field-choices-enum-types>` directly instead of
   requiring expansion with the ``choices`` attribute.
 
+* Added ``django.models.fields.GeneratedField``.
```
1. link to the appropriate docs
2. summarize the feature

Without these, the release note is not going to surface your hard work and make it easy for others to discover :)

```suggestion
* The new :class:`~django.db.models.fields.GeneratedField` allows creation of
  database generated columns.
```

pauloxnet replied:
LGTM

adamchainz commented on docs/ref/models/fields.txt:
```diff
@@ -1150,6 +1150,44 @@ when :attr:`~django.forms.Field.localize` is ``False`` or
+
+.. class:: GeneratedField(expression, output_field=None, db_persist=True, **kwargs)
+
+A field which is computed based on other fields in the model. Uses the database's
+``GENERATED ALWAYS AS`` syntax. Similar to a database view, this field is managed
+and updated by the database itself.
```
Slight reordering of details, since SQL is a bit more of an implementation detail.

```suggestion
A field which is computed based on other fields in the model. Similar to a
database view, this field is managed and updated by the database itself. Uses
the ``GENERATED`` SQL syntax.
```

Also could we add an example after this paragraph? Just a basic use case, maybe from the tests, would really help users see what this does.

adamchainz commented on docs/ref/models/fields.txt:
```diff
@@ -1150,6 +1150,44 @@ when :attr:`~django.forms.Field.localize` is ``False`` or
+
+.. attribute:: GeneratedField.output_field
+
+    An optional :class:`Field` used to explicitly define the model's underlying
+    column type. This can be used to customize the field's DB collation or other
+    field-specific properties.
+
+    By default, this field is derived from the given ``expression``.
```
Tried to tighten this up. "Properties" isn't quite the right word - that can lead to confusion with `@property`. "attribute" is what we really talk about in Python.

```suggestion
    An optional :class:`Field` to define the field’s data type. This can be
    used to customize attributes like the field's collation. By default, the
    output field is derived from ``expression``.
```

jpnauta commented:
Thanks all for the feedback! I believe I implemented addressed all your comments.

I haven't implemented Oracle because I haven't found any way to run it locally. Any tips? I'd be happy to implement the features, but I currently have no way of verifying the changes work as expected.

piw2266 commented:
하늘2023. 3. 29. 12:00에 Jeremy Nauta ***@***.***>님이 작성:
Thanks all for the feedback! I believe I implemented addressed all your comments.
I haven't implemented Oracle because I haven't found any way to run it locally. Any tips? I'd be happy to implement the features, but I currently have no way of verifying the changes work as expected.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

piw2266 commented:
하늘2023. 3. 29. 12:44에 이 포도 ***@***.***>님이 작성:하늘2023. 3. 29. 12:00에 Jeremy Nauta ***@***.***>님이 작성:
Thanks all for the feedback! I believe I implemented addressed all your comments.
I haven't implemented Oracle because I haven't found any way to run it locally. Any tips? I'd be happy to implement the features, but I currently have no way of verifying the changes work as expected.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

piw2266 commented:
하늘2023. 3. 29. 12:45에 이 포도 ***@***.***>님이 작성:하늘2023. 3. 29. 12:44에 이 포도 ***@***.***>님이 작성:하늘2023. 3. 29. 12:00에 Jeremy Nauta ***@***.***>님이 작성:
Thanks all for the feedback! I believe I implemented addressed all your comments.
I haven't implemented Oracle because I haven't found any way to run it locally. Any tips? I'd be happy to implement the features, but I currently have no way of verifying the changes work as expected.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

jpnauta commented:
@felixxm @pauloxnet @adamchainz any comments on this P/R, or tips for implementing Oracle?

pauloxnet commented:
> @felixxm @pauloxnet @adamchainz any comments on this P/R, or tips for implementing Oracle?

LGTM

I've no experience with Oracle , so I can't help you on that point :-(

adamchainz commented:
Personally I'm fine with us leaving out Oracle support for new features like this, if it's troublesome. I am not sure we should be chasing all of its quirks for all features at this point. [Only 2%](https://lp.jetbrains.com/django-developer-survey-2022/#text-403) of survey respondents use it.

That said, it may be worth a try. The django-docker-box project helps you run multiple database backends - it has [instructions for setting up Oracle](https://github.com/django/django-docker-box#oracle). Try that if you have the time and energy.

Otherwise @felixxm loves Oracle (😜) and may be able to help.

pauloxnet commented:
> Personally I'm fine with us leaving out Oracle support for new features like this, if it's troublesome.

I agree with you.

jpnauta commented:
I attempted to use [django-docker-box](https://github.com/django/django-docker-box/) but ran into issue after issue. Seems to be a [known issue](https://github.com/django/django-docker-box/issues/4). I got close to running it using `truevoly/oracle-12c`, only to realize that Oracle 19 is required. 

Unless anyone can help implement this feature in Oracle, I'm going to consider this P/R ready to merge!

FYI @felixxm

jpnauta commented:
Thanks for the info @LilyFoote! Interested in working together to implement Oracle support? If you're willing to review this P/R and QA it for Oracle, then I'd be happy to take a "best guess" at what the implementation would be. How does that sound?

LilyFoote commented:
@jpnauta Actually, I should have a PR with a few commits for you soon adding a first draft of Oracle support.

adamchainz commented:
Hero!

jpnauta commented:
Thanks @LilyFoote for implementing Oracle support! I have merged the latest from `main` now, I believe this is ready to go.

@felixxm @pauloxnet are we able to get this P/R merged so we can move forward towards Lily's P/R?

pauloxnet commented:
This PR LGTM

pauloxnet commented:
@jpnauta can you rebase this PR so that can be easier to review for others?

jpnauta commented:
@pauloxnet done! FYI @LilyFoote you will need to rebase as well 😄

LilyFoote commented:
@jpnauta rebased.

schinckel commented:
I'm pretty late to the party, but I only just came across this today.

> a_plus_b = models.GeneratedField(expression=F("a") + F("b"))

Will this field update within the context of a model instance if the values of a and/or b are changed?

jpnauta commented:
> I'm pretty late to the party, but I only just came across this today.
> 
> > a_plus_b = models.GeneratedField(expression=F("a") + F("b"))
> 
> Will this field update within the context of a model instance if the values of a and/or b are changed?

Yes @schinckel, there are a few tests in this P/R that do exactly that! Regardless of when/how the model's fields are changed, the generated field will always be `a` plus `b`.

schinckel commented:
Thanks for the reply. I've looked at the tests, and it seems to me that a refresh from the database is required to trigger the value to update.

(I'm asking because I implemented https://django-shared-property.readthedocs.io/en/latest/ specifically to handle hybrid properties. I'm going to look deeper into how you implemented this to see if I can make live-updating-properties work on top of this PR).

schinckel commented:
Yeah, that's not quite what I meant.

This was more about an update to a dependent column affecting the generated column's value without having _any_ database access required.

jpnauta commented:
@schinckel gotcha - in that case, no, this feature requires a DB update for the feature to work

schinckel commented:
Coolio. I think I can still make my stuff (which already works) build on this PR when it's merged anyway. I implemented allowing related objects' fields as part of the interface, but I think it's probably "better" to restrict to only local fields.

felixxm commented on django/db/models/fields/generated.py:
```diff
@@ -0,0 +1,137 @@
+        self._output_field = output_field
+        self.db_persist = db_persist
+
+        assert "editable" not in kwargs, "Cannot set `editable` for GeneratedField"
+        assert "blank" not in kwargs, "Cannot set `blank` for GeneratedField"
+        assert "default" not in kwargs, "Cannot set `default` for GeneratedField"
```
This is a user-facing error so we shouldn't use `assert` (check out ticket-32508).

felixxm commented:
Closing in favor or #16860.
